### ISSUE TITLE
Create API access library in Python

### ISSUE THREAD

**Issue by rengolin** [mr-provisioner/mr-provisioner#121]
Right now, the Ansible roles access the API via curl, which works fine, but it's cumbersome and makes error checking repetitive and error-prone. Furthermore, anything more complicated than what Ansible can do ends up being a script (hopefully Python, not Bash or Perl), which then can't easily use the existing Ansible roles for interfacing with Mr-Provisioner.

More importantly, we can't do a lot of testing via the API from the UI and it's cumbersome to do that via Ansible. Without tests, it's hard to know what changes to the API create subtle errors or unintended consequences. An official client would be perfect for generating good test coverage that could be done from anywhere, even as a pull request trigger.

Our Jenkins benchmark job, for instance, is in Python, but needs to get info from Ansible (machine info) and then pass it along all the way down to the script. Clearly, it should be the other way round.

I propose we create a standard Python module/package, which fully implements the API, and has to be updated every time the API changes.

This package shall be used for the following purposes:
 1. Writing unit tests for the API to guarantee stability and pull request validation
 2. Used by Python scripts to interface with Mr-Provisioner directly, which also help creating further modules/packages relying on Mr-Provisioner
 3. Used by Ansible as a Python invocation instead of hand-crafted curl calls

This would make the product more stable, usable and easier to debug.

@niedbalski @BaptisteGerondeau @danrue

**Comment by gemagomez** [comment]
> More importantly, we can't do a lot of testing via the API from the UI and it's cumbersome to do that via Ansible. Without tests, it's hard to know what changes to the API create subtle errors or unintended consequences. An official client would be perfect for generating good test coverage that could be done from anywhere, even as a pull request trigger.

We have API tests that fail when compatibility is broken and get run on PRs: https://github.com/mr-provisioner/mr-provisioner/tree/master/tests

> Our Jenkins benchmark job, for instance, is in Python, but needs to get info from Ansible (machine info) and then pass it along all the way down to the script. Clearly, it should be the other way round.

I don't understand what this means.

> I propose we create a standard Python module/package, which fully implements the API, and has to be updated every time the API changes.

This sounds like a useful feature.

> This package shall be used for the following purposes:
> 
> Writing unit tests for the API to guarantee stability and pull request validation

It should have tests, the same way the API and any other core functionality should have tests.

> Used by Ansible as a Python invocation instead of hand-crafted curl calls

It would help client applications written in python. The API will still be the official point of contact/compatibility within the app, though.

**Comment by rengolin** [comment]
It seems we're talking about two different things. Your tests use the objects directly, not the REST API, and those are two different things. The objects can change their behaviour internally, as long as the API is stable (yields the same results), it doesn't need to change version, update clients, etc.

What I'm talking about is the public API, the one that the website and @danrue's Python wrapper uses, and the one this new Python client should use. My point is to make sure all of them have the same entry point (the public REST API), so that when you test one (ex. the client), you're testing all.

This test won't be as simple as the ones you have now. You'd need a working provisioner to test the API calls on, which may not be feasible on every pull request, but one that should be executed on every merge, to make sure functionality wise, we're still behaving as expected.

The Python client would, in addition, be a useful way to import this functionality into a Python program, much like @danrue's, eventually replacing it altogether. So, his Ansible roles will just use the client directly, and so will our Ansible / Python scripts.

Our Jenkins benchmark job is written in Python, and right now we have to add Ansible wrappers *just* so we can query the machine's IP address. It would be good to be able to get the address from inside the Python [harness](https://github.com/Linaro/benchmark_harness) we already have.

**Comment by gemagomez** [comment]
> It seems we're talking about two different things. Your tests use the objects directly, not the REST API, and those are two different things. The objects can change their behaviour internally, as long as the API is stable (yields the same results), it doesn't need to change version, update clients, etc.

The tests we have test the rest API and verify the results using the objects, this is standard API testing ( (i.e. https://github.com/mr-provisioner/mr-provisioner/blob/master/tests/api/test_auth.py).

**Comment by rengolin** [comment]
That's... 3 tests.

**Comment by gemagomez** [comment]
It was an example, I assumed you could navigate the code by yourself, but here you go:

https://github.com/mr-provisioner/mr-provisioner/blob/master/tests/api/test_image.py
https://github.com/mr-provisioner/mr-provisioner/blob/master/tests/api/test_machine.py
https://github.com/mr-provisioner/mr-provisioner/blob/master/tests/api/test_preseed.py

**Comment by rengolin** [comment]
Right, so we have tests. Good. Then we create a Python client that does exactly the same thing, and then how do we test that? Well, if we test one, we test all. If the Python client is official, and we test the API via the client, we don't need to test the API via any other means.

If we have two sets of identical tests, they will eventually diverge with time. That's all I'm saying.

**Comment by gemagomez** [comment]
> Right, so we have tests. Good. Then we create a Python client that does exactly the same thing, and then how do we test that? Well, if we test one, we test all. If the Python client is official, and we test the API via the client, we don't need to test the API via any other means.

If/when we have a python client, that client will need tests. Even if they are testing exactly the same thing the API tests are testing. This would be to test the client, not the API.

> If we have two sets of identical tests, they will eventually diverge with time. That's all I'm saying.

They will not be identical. One set of tests will target the API, one set will target the client. They will not diverge with time if the client keeps up with the API. The API, however, needs to remain compatible with any client out there, python or otherwise, so the tests are not redundant. The client does not superseed the API.

**Comment by rengolin** [comment]
Sounds good.

**Comment by LGraemeG** [comment]
>  If the Python client is official, and we test the API via the client, we don't need to test the API via any other means.

What this means in reality is you end up with holes in your testing because you client makes assumptions about the API which may or may not be true in all cases.

**Comment by danrue** [comment]
I don't think a blessed python client is necessary for a REST api. Maybe if there were a need for a cli utility, someone could make one and then the ansible role could use its libraries. There just isn't enough complexity in sending a POST, PUT, or GET to a url to justify another layer of abstraction. Granted, there are a few functions which might benefit from being in a distributed library ([these lookup functions](https://github.com/Linaro/ansible-role-mr-provisioner/blob/master/library/mr_provisioner_machine_provision.py#L132-L181) come to mind) - but those would be better addressed in the API itself to make them unnecessary.

I think this issue can be distilled into 'ansible-role-mr-provisioner needs local tests'. And I heartily concur.

**Comment by rengolin** [comment]
This is a matter of:
1. Commoning up the expected client logic somewhere
2. Commoning up error handling in a single place
3. Moving an actual Python client outside of an Ansible repository

REST not being complex doesn't even factor in.

**Comment by danrue** [comment]
@rengolin - if you want to make a well tested python client, I'll gladly use it, and I would accept a pull request to convert the ansible role to use it as well.

**Comment by rengolin** [comment]
@danrue https://github.com/Linaro/mr-provisioner-client
